stackrox/stackrox
# ROX-18686: Add missing resourceAccessRequirements in routePaths

## Description

1. Replace implicit **and** relationship for `resourceAccessRequirements` with predicates functions:
    * `everyResource` for **and** relationship
    * `someResource` for **or** relationship

2. Add conditional rendering in main Dashboard
    * `ViolationsByPolicySeverity` only if has read access for Alert
    * `ImagesAtMostRisk` only if has read access for Image
    * `DeploymentsAtMostRisk` only if has read access for Deployment
    * `AgingImages` only if has read access for Image
    * `ViolationsByPolicyCategory` only if has read access for Alert
    * `ComplianceLevelsByStandard` only if has read access for Compliance

3. Add conditional rendering in Risk
    * **View Deployment in Network Graph** on if has requirements for `networkPath`
    * **Process Discovery** only if read access for Alert

4. Add conditional rendering in Clusters
    * Render **Manage Tokens** and **Upgrade** only if write access for Integration
    * Render **Delete**, and **Install cluster** only if write access for Cluster
    * Render **Next** only if write access for Cluster

5. Add conditional rendering in Policy Management
   * Render **Bulk actions** only if write access for WorkflowAdministration
    * Render **Create policy** and **Import policy** only if write access for WorkflowAdministration
    * Render **Create category** only if write access for WorkflowAdministration

6. Add conditional rendering in System Health
    * Render cluster health only if read access for Cluster
    * Render vulnerability definitions health only if read access for Administration
    * Render integration health only if read access for Integration
    * Render declarative configuration health only if read access for Integration

### Residue

1. Conditional requests and rendering for `SummaryCounts` on main dashboard.
2. Conditional rendering for **Create Policy** on Risk.
3. Vulnerability Management 1.0
4. Configuration Management
5. src/Containers/Clusters/Components/CredentialExpirationWidget.tsx
    * `ManageTokensButton` needs conditional rendering only if write access for Integration
    * `CredentialInteraction` needs conditional rendering only if write access for Administration
6. Delete `AccessControlNoPermission` element
7. Dig deeper to find more interactions that require READ_WRITE_ACCESS
8. Write integration tests for conditional rendering

## Checklist

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.css: 609 = 3759098 - 3758489
        total: 1353 = 9920560 - 9919207
    * `ls -al build/static/js/*.js | wc -l`
        0 = 80 - 80 files
3. `yarn start` in ui

### Manual testing

**Thanks to Van** for explaining how to test with specific permissions using Admin in one window and an auth provider in another window.

1. Visit /main/dashboard with NO_ACCESS for all resources
    * See residue 1: request failures for summary counts.
    * See conditional rendering for widgets, which do not have request failures, because each takes responsibility for its own data.
        ![dashboard](https://github.com/stackrox/stackrox/assets/11862657/e8fe5ded-0822-4f3f-9bb2-44a0f8f45bca)

2. Visit /main/risk with READ_ACCESS for Deployment and DeploymentExtension
    * See residue 2: lack of conditional rendering for **Create Policy**
    * See conditional rendering for **View Deployment in Network Graph** on **Deployment Details** tab
    * See conditional rendering for **Process Discovery** tab
        ![risk](https://github.com/stackrox/stackrox/assets/11862657/aba7bd97-6c37-43b8-b8aa-0963c5512ba5)

3. Visit /main/clusters
    * See conditional rendering for **Manage Tokens** and **Upgrade**
    * See conditional rendering for **Delete**, and **Install cluster**
        ![clusters](https://github.com/stackrox/stackrox/assets/11862657/a7cd66b6-653e-475b-9a6c-869475bf7659)

4. Click a row
    * See conditional rendering for **Next**
        ![cluster](https://github.com/stackrox/stackrox/assets/11862657/a6afb663-269a-45a5-8648-97557715da25)

5. Visit /main/policy-management/policies
    * See conditional rendering for **Bulk actions** **Create policy** **Import policy**
        ![policies](https://github.com/stackrox/stackrox/assets/11862657/e79d8433-ec74-4652-a622-3e2437908e87)

6. Click **Policy categories**
    * See conditional rendering for **Create category**
        ![policy-categories](https://github.com/stackrox/stackrox/assets/11862657/c60e328c-0798-4e20-882d-691eaffe7ab2)

7. Visit /main/system-health and I determined conditional rendering by investigating request failures
